packages/logging/src/metadata.js

```diff
@@ -0,0 +1,297 @@
+        // but the metadata service, on non-200 responses, often returns a full
+        // HTML page which wouldn't much help as an error message anyway.
+        var err = new Error(
+          'Request to get project id from metadata service failed with code: ' +
+          resp.statusCode
+        );
+        self.metadataContactError_ = err;
+        // Remove the data listener after emitting the failure event. This will
+        // ensure that the end event listeners condition of an empty project id
+        // is met, thereby ensuring that only the failure event is emitted.
+        self.metadataRequestStream_.emit(REQUEST_EVENTS.FAILURE, err);
+        self.metadataRequestStream_.removeAllListeners();
+      }
+    })
+    .once('end', function() {
+      if (!is.empty(projectId)) {
+        // The project id was set by the data event, cache it on the instance
+        // and emit the success event to all applicable listeners which have
+        // queued.
+        self.metadataProjectId_ = projectId;
+        self.metadataRequestStream_.emit(REQUEST_EVENTS.SUCCESS,
+          self.metadataProjectId_);
+      } else {
+        // The stream was ended before a response could be gathered
+        self.metadataContactError_ = new Error('Stream was prematurely ended');
+        self.metadataRequestStream_.emit(REQUEST_EVENTS.FAILURE,
+          self.metadataContactError_);
+      }
+      self.metadataRequestStream_.removeAllListeners();
+    });
+};
+
+/**
+ * Returns the environment variable sometimes containing the project id
+ *
+ * @private
+ * @returns {Any} - the env property GCLOUD_PROJECT
+ */
+Metadata.prototype.processProjectId_ = function() {
+  return process.env.GCLOUD_PROJECT;
+};
+
+/**
+ * Called to determine whether the instance is a cloud function
+ *
+ * @private
+ * @returns {Boolean} - Whether or not the env property FUNCTION_NAME is empty
+ */
+Metadata.prototype.isCloudFunction_ = function() {
+  return !is.empty(process.env.FUNCTION_NAME);
+};
+
+/**
+ * Called to determine whether the instance is on app engine
+ *
+ * @private
+ * @returns {Boolean} - Whether or not the env property GAE_SERVICE is empty
+ */
+Metadata.prototype.isAppEngine_ = function() {
+  return !is.empty(process.env.GAE_SERVICE);
+};
+
+/**
+ * Called to create a descriptor for Google Compute Engine
+ *
+ * @private
+ * @param {String} projectId - the project id
+ * @returns {Object}
+ */
+Metadata.prototype.produceGCEDescriptor_ = function(projectId) {
+  return {
+    type: 'gce_instance',
+    labels: {
+      project_id: projectId
+    }
+  };
+};
+
+/**
+ * Called to create a descriptor for Google App Engine
+ *
+ * @private
+ * @param {String} projectId - the project id
+ * @returns {Object}
+ */
+Metadata.prototype.produceGAEDescriptor_ = function(projectId) {
+  return {
+    type: 'gae_app',
+    labels: {
+      project_id: projectId,
+      module_id: process.env.GAE_SERVICE,
+      version_id: process.env.GAE_VERSION
```

Comment:
ManagedVMs are the slightly older variant of AppEngine Flexible – you get this variant when you have `vm: true` in your `app.yaml`. As far as I know `GAE_SERVICE` is not defined on ManagedVMs. From the user's point of view both of these are 'AppEngine' and we should magically discover & configure either. /cc @JustinBeckwith in case I got these details wrong.